lib/ts/recipe/totp/prebuiltui.tsx

```diff
@@ -0,0 +1,102 @@
+
+    // Static methods
+    static getInstanceOrInitAndGetInstance(): TOTPPreBuiltUI {
+        if (TOTPPreBuiltUI.instance === undefined) {
+            const recipeInstance = TOTPRecipe.getInstanceOrThrow();
+            TOTPPreBuiltUI.instance = new TOTPPreBuiltUI(recipeInstance);
+        }
+
+        return TOTPPreBuiltUI.instance;
+    }
+    static getFeatures(
+        useComponentOverrides: () => GenericComponentOverrideMap<any> = useRecipeComponentOverrideContext
+    ): RecipeFeatureComponentMap {
+        return TOTPPreBuiltUI.getInstanceOrInitAndGetInstance().getFeatures(useComponentOverrides);
+    }
+    static getFeatureComponent(
+        componentName: "mfaTOTP",
+        props: any,
+        useComponentOverrides: () => GenericComponentOverrideMap<any> = useRecipeComponentOverrideContext
+    ): JSX.Element {
+        return TOTPPreBuiltUI.getInstanceOrInitAndGetInstance().getFeatureComponent(
+            componentName,
+            props,
+            useComponentOverrides
+        );
+    }
+
+    // Instance methods
+    getFeatures = (
+        useComponentOverrides: () => GenericComponentOverrideMap<any> = useRecipeComponentOverrideContext
+    ): RecipeFeatureComponentMap => {
+        const features: RecipeFeatureComponentMap = {};
+        if (this.recipeInstance.config.totpMFAScreen.disableDefaultUI !== true) {
+            const normalisedFullPath = this.recipeInstance.config.appInfo.websiteBasePath.appendPath(
+                new NormalisedURLPath(DEFAULT_TOTP_PATH)
+            );
+            features[normalisedFullPath.getAsStringDangerous()] = {
+                matches: matchRecipeIdUsingQueryParams(this.recipeInstance.config.recipeId),
+                component: (props: any) => this.getFeatureComponent("mfaTOTP", props, useComponentOverrides),
+                recipeID: TOTPRecipe.RECIPE_ID,
+            };
+        }
+        return features;
+    };
+    getFeatureComponent = (
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        _: "mfaTOTP",
+        props: any,
+        useComponentOverrides: () => GenericComponentOverrideMap<any> = useRecipeComponentOverrideContext
+    ): JSX.Element => {
+        return (
+            <UserContextWrapper userContext={props.userContext}>
+                <SessionAuth overrideGlobalClaimValidators={() => []}>
+                    <MFATOTPFeature
+                        recipe={this.recipeInstance}
+                        useComponentOverrides={useComponentOverrides}
+                        {...props}
+                    />
+                </SessionAuth>
+            </UserContextWrapper>
+        );
+    };
+
+    // For tests
+    static reset(): void {
+        if (!isTest()) {
+            return;
+        }
+
+        TOTPPreBuiltUI.instance = undefined;
+        return;
+    }
+
+    static MFATOTP = (props?: any): JSX.Element =>
+        TOTPPreBuiltUI.getInstanceOrInitAndGetInstance().getFeatureComponent("mfaTOTP", props);
```

Comment:
this one component will be responsible for creating + verifying a device and doing the totp challenge? Should we expose two components instead?

If we expose just one component for both, and the user wants to embed this, and show creating device UI in one case vs enter totp in another case, how would they do this?

Reply:
Well, we do want to have a single path handling both things, which would mean a single feature component.
If someone wants to re-use our UI, but not the logic which auto-creates the device if necessary they can just take our theme components. Partially re-using the feature component can be done, but it's not trivial.